### PR TITLE
cli/demo: various fixes

### LINKED ISSUE #102257: acceptance: TestDockerCLI_test_demo_node_cmds failed
acceptance.TestDockerCLI_test_demo_node_cmds [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_Acceptance/9796707?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_Acceptance/9796707?buildTab=artifacts#/) on master @ [6563de01d12656eccf28f9dd1722e0052a75d512](https://github.com/cockroachdb/cockroach/commits/6563de01d12656eccf28f9dd1722e0052a75d512):

```
=== RUN   TestDockerCLI_test_demo_node_cmds
    test_log_scope.go:161: test logs captured to: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/_tmp/0f973bfae5165b427f0ee0932d9c6ec9/logTestDockerCLI_test_demo_node_cmds47575177
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestDockerCLI_test_demo_node_cmds
    generated_cli_test.go:131: -- test log scope end --
test logs left over in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/_tmp/0f973bfae5165b427f0ee0932d9c6ec9/logTestDockerCLI_test_demo_node_cmds47575177
--- FAIL: TestDockerCLI_test_demo_node_cmds (58.25s)
=== RUN   TestDockerCLI_test_demo_node_cmds/test_demo_node_cmds.tcl
=== CONT  TestDockerCLI_test_demo_node_cmds/test_demo_node_cmds.tcl
    util_docker.go:254: non-zero exit code: 1
    --- FAIL: TestDockerCLI_test_demo_node_cmds/test_demo_node_cmds.tcl (44.54s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-sessions
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDockerCLI_test_demo_node_cmds.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-27369
Epic: CRDB-28893

### LINKED ISSUE #107888: multitenant: `\demo add` is broken in multi-tenant mode
**Describe the problem**

The `\demo add` statement which adds nodes to a demo cluster is broken when demo runs in multi-tenant mode.

**To Reproduce**

* `cockroach demo --nodes 9`
* `\demo add "region=us-central1"`
* `\demo ls`

```
Application tenant:
*
* ERROR: a panic has occurred!
* runtime error: index out of range [9] with length 9
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:890
*   | runtime.goPanicIndex
*   | 	GOROOT/src/runtime/panic.go:113
*   | github.com/cockroachdb/cockroach/pkg/cli/democluster.(*transientCluster).ListDemoNodes
*   | 	github.com/cockroachdb/cockroach/pkg/cli/democluster/demo_cluster.go:1902
*   | github.com/cockroachdb/cockroach/pkg/cli/clisqlshell.(*cliState).handleDemo
*   | 	github.com/cockroachdb/cockroach/pkg/cli/clisqlshell/sql.go:696
*   | github.com/cockroachdb/cockroach/pkg/cli/clisqlshell.(*cliState).doHandleCliCmd
*   | 	github.com/cockroachdb/cockroach/pkg/cli/clisqlshell/sql.go:1579
*   | github.com/cockroachdb/cockroach/pkg/cli/clisqlshell.(*cliState).doRunShell
*   | 	github.com/cockroachdb/cockroach/pkg/cli/clisqlshell/sql.go:2326
*   | github.com/cockroachdb/cockroach/pkg/cli/clisqlshell.(*cliState).RunInteractive
*   | 	github.com/cockroachdb/cockroach/pkg/cli/clisqlshell/sql.go:2275
*   | github.com/cockroachdb/cockroach/pkg/cli/clisqlcfg.(*Context).Run
*   | 	github.com/cockroachdb/cockroach/pkg/cli/clisqlcfg/context.go:225
*   | github.com/cockroachdb/cockroach/pkg/cli.runDemoInternal
*   | 	github.com/cockroachdb/cockroach/pkg/cli/demo.go:383
*   | github.com/cockroachdb/cockroach/pkg/cli.runDemo
*   | 	github.com/cockroachdb/cockroach/pkg/cli/demo.go:200
*   | github.com/cockroachdb/cockroach/pkg/cli.init.5.func1
*   | 	github.com/cockroachdb/cockroach/pkg/cli/demo.go:59
*   | github.com/cockroachdb/cockroach/pkg/cli/clierrorplus.MaybeDecorateError.func1
*   | 	github.com/cockroachdb/cockroach/pkg/cli/clierrorplus/decorate_error.go:72
*   | github.com/spf13/cobra.(*Command).execute
*   | 	github.com/spf13/cobra/external/com_github_spf13_cobra/command.go:856
*   | github.com/spf13/cobra.(*Command).ExecuteC
*   | 	github.com/spf13/cobra/external/com_github_spf13_cobra/command.go:974
*   | github.com/spf13/cobra.(*Command).Execute
*   | 	github.com/spf13/cobra/external/com_github_spf13_cobra/command.go:902
*   | github.com/cockroachdb/cockroach/pkg/cli.Run
*   | 	github.com/cockroachdb/cockroach/pkg/cli/cli.go:300
*   | github.com/cockroachdb/cockroach/pkg/cli.doMain
*   | 	github.com/cockroachdb/cockroach/pkg/cli/cli.go:143
*   | github.com/cockroachdb/cockroach/pkg/cli.Main
*   | 	github.com/cockroachdb/cockroach/pkg/cli/cli.go:68
*   | main.main
*   | 	main/pkg/cmd/cockroach/main.go:26
*   | runtime.main
*   | 	GOROOT/src/runtime/proc.go:250
*   | runtime.goexit
*   | 	GOROOT/src/runtime/asm_arm64.s:1172
* Wraps: (2) runtime error: index out of range [9] with length 9
* Error types: (1) *withstack.withStack (2) runtime.boundsError
*
panic: runtime error: index out of range [9] with length 9 [recovered]
	panic: runtime error: index out of range [9] with length 9 [recovered]
	panic: runtime error: index out of range [9] with length 9

goroutine 1 [running]:
github.com/cockroachdb/cockroach/pkg/util/log/logcrash.RecoverAndReportPanic({0x109ab68c0, 0x140001a8010}, 0x140283e8c88?)
	github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:132 +0x74
panic({0x1096a9820, 0x1403f187248})
	GOROOT/src/runtime/panic.go:884 +0x204
github.com/cockroachdb/cockroach/pkg/cli/democluster.(*transientCluster).Close(0x14002f20000, {0x109ab68c0?, 0x140001a8010})
	github.com/cockroachdb/cockroach/pkg/cli/democluster/demo_cluster.go:968 +0x1cc
panic({0x1096a9820, 0x1403f187248})
	GOROOT/src/runtime/panic.go:890 +0x258
github.com/cockroachdb/cockroach/pkg/cli/democluster.(*transientCluster).ListDemoNodes(0x14002f20000, {0x109a70aa0, 0x140001c0008}, {0x109a70aa0, 0x140005aa060}, 0x0, 0x1)
	github.com/cockroachdb/cockroach/pkg/cli/democluster/demo_cluster.go:1902 +0x9d8
github.com/cockroachdb/cockroach/pkg/cli/clisqlshell.(*cliState).handleDemo(0x1401befe940?, {0x14024234050?, 0x10748a028?, 0x1?}, 0x140627271f8?, 0x0?)
	github.com/cockroachdb/cockroach/pkg/cli/clisqlshell/sql.go:696 +0x84
github.com/cockroachdb/cockroach/pkg/cli/clisqlshell.(*cliState).doHandleCliCmd(0x1401cbd1b80, 0x2, 0x14028f29308?)
	github.com/cockroachdb/cockroach/pkg/cli/clisqlshell/sql.go:1579 +0xa50
github.com/cockroachdb/cockroach/pkg/cli/clisqlshell.(*cliState).doRunShell(0x1401cbd1b80, 0x119ac5cb8?, 0x10f8bcf18?, 0xd0?, 0x140005aa060)
	github.com/cockroachdb/cockroach/pkg/cli/clisqlshell/sql.go:2326 +0x2fc
github.com/cockroachdb/cockroach/pkg/cli/clisqlshell.(*cliState).RunInteractive(0x14000e431d0?, 0x109b25260?, 0x14011ecf520?, 0x140130b0870?)
	github.com/cockroachdb/cockroach/pkg/cli/clisqlshell/sql.go:2275 +0x5c
github.com/cockroachdb/cockroach/pkg/cli/clisqlcfg.(*Context).Run(0x14000e431d0, {0x109ab68c0, 0x140001a8010}, {0x109b25260?, 0x14011ecf520})
	github.com/cockroachdb/cockroach/pkg/cli/clisqlcfg/context.go:225 +0x2ec
github.com/cockroachdb/cockroach/pkg/cli.runDemoInternal(0x10dab3020, {0x0, 0x0}, 0x109a08a80)
	github.com/cockroachdb/cockroach/pkg/cli/demo.go:383 +0xc40
github.com/cockroachdb/cockroach/pkg/cli.runDemo(...)
	github.com/cockroachdb/cockroach/pkg/cli/demo.go:200
github.com/cockroachdb/cockroach/pkg/cli.init.5.func1(0x9?, {0x1400099da38?, 0x1?, 0xc?})
	github.com/cockroachdb/cockroach/pkg/cli/demo.go:59 +0x2c
github.com/cockroachdb/cockroach/pkg/cli/clierrorplus.MaybeDecorateError.func1(0x10dab3020?, {0x1400129e1e0?, 0x2?, 0x2?})
	github.com/cockroachdb/cockroach/pkg/cli/clierrorplus/decorate_error.go:72 +0x34
github.com/spf13/cobra.(*Command).execute(0x10dab3020, {0x1400129e1c0, 0x2, 0x2})
	github.com/spf13/cobra/external/com_github_spf13_cobra/command.go:856 +0x4dc
github.com/spf13/cobra.(*Command).ExecuteC(0x10daab0a0)
	github.com/spf13/cobra/external/com_github_spf13_cobra/command.go:974 +0x360
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra/external/com_github_spf13_cobra/command.go:902
github.com/cockroachdb/cockroach/pkg/cli.Run(...)
	github.com/cockroachdb/cockroach/pkg/cli/cli.go:300
github.com/cockroachdb/cockroach/pkg/cli.doMain(0x10dab3020, {0x1400340401a, 0x4})
	github.com/cockroachdb/cockroach/pkg/cli/cli.go:143 +0x1cc
github.com/cockroachdb/cockroach/pkg/cli.Main()
	github.com/cockroachdb/cockroach/pkg/cli/cli.go:68 +0x194
main.main()
	main/pkg/cmd/cockroach/main.go:26 +0x1c
```

**Expected behavior**
No panic.  Lists all nodes (including the newly added 10th node)

Jira issue: CRDB-30244
Epic: CRDB-26691

### LINKED ISSUE #108331: acceptance: flake in TestDockerCLI_test_demo_node_cmds
Seen [here](https://teamcity.cockroachdb.com/viewLog.html?buildId=11226084&buildTypeId=Cockroach_BazelEssentialCi) on an unrelated PR.
```
.230807 21:30:21.046279721 EXPECT TEST: TIMEOUT WAITING FOR "node 6 has been added with locality "region=ca-central,zone=a""
system(echo; echo $(date '+.%y%m%d %H:%M:%S.%N') EXPECT TEST: 'TIMEOUT WAITING FOR "node 6 has been added with locality "region=ca-central,zone=a""' | tee -a logs/expect-cmd.log) = 2
system(find logs -exec chmod a+rw '{}' \; || true) = 2
non-zero exit code: 1
```
Seems like a dup of #102257, cc @rafiss 

Jira issue: CRDB-30427
Epic: CRDB-28893

### LINKED ISSUE #108563: cli/demo: the `\demo recommission` command has become non-sensical since v20.2
In v20.2 we have introduced the 2-step state transition decommissioning -> decommissioned.
once a node is in the "decommissioned" state the state is permanent -- i.e. it cannot be recommissioned.

Meanwhile, the `\demo decommission` command performs _both_ state changes i.e. it has become unrecoverable since v20.2.

This means that `\demo recommission` simply cannot exist any more. It should be removed.

Epic: CRDB-28893

Jira issue: CRDB-30522

### PR DESCRIPTION
**Pull request by knz**
See individual commits for details.

Informs (and will fix after backport) #96239.
Fixes #102257.
Fixes #108331.
Fixes #108563.
Fixes #107888.

Epic: CRDB-28893

### CODE REVIEW

File: pkg/cli/interactive_tests/test_demo_multitenant.tcl.disabled
Comment (rafiss) [inline pkg/cli/interactive_tests/test_demo_multitenant.tcl.disabled:1]: nit: the file needs to be renamed

**Review by rafiss — APPROVED** [review]
very nice! thanks for discovering all this and improving it. my only comments are about comments + commit messages

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108566#-:-Nb_UTRI70DCIAW-z1wr:bjx0hvj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @yuzefovich)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108566#-Nb_TO4VAAso5YvDbyJM):*
nit: consider mentioning a one sentence summary of what the fix was. most of the change seems to be just moving code from one function to another. was the problem the missing call in `startServerInternal`? 

does this first commit also fix the "timeout" issue in https://github.com/cockroachdb/cockroach/issues/108331
___
*[`pkg/cli/democluster/demo_cluster.go` line 503 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108566#-Nb_Twmv4R02A10f1WoL:-Nb_Twmv4R02A10f1WoM:b-xnovy0) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fc21d9d972a9933aca20ed6c5760d8b4f6e205c/pkg/cli/democluster/demo_cluster.go#L503)):*
> ```Go
> }
> 
> func (c *transientCluster) startTenantService(
> ```

nit: this function could use a comment

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108566)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108566#-Nb_TO4VAAso5YvDbyJM):*
> nit: consider mentioning a one sentence summary of what the fix was. 

Done.

> most of the change seems to be just moving code from one function to another. was the problem the missing call in `startServerInternal`?

Yes.

> does this first commit also fix the "timeout" issue in [ acceptance: flake in TestDockerCLI\_test\_demo\_node\_cmds #108331](https://github.com/cockroachdb/cockroach/issues/108331)

This was due to the broken decommission (`\demo decommission`  needs a drain/shutdown before the decommission proper). Addressed in r5. I also added a release note there.
___
*[`pkg/cli/democluster/demo_cluster.go` line 503 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108566#-Nb_Twmv4R02A10f1WoL:-Nba3hH-06ZGjV-DXvX8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fc21d9d972a9933aca20ed6c5760d8b4f6e205c/pkg/cli/democluster/demo_cluster.go#L503)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: this function could use a comment
</blockquote></details>

Done.

**Comment by knz** [comment]
TFYR!

bors r=rafiss

**Comment by knz** [comment]
bors r=rafiss

**Comment by knz** [comment]
backport here #108631

**Review by yuzefovich — COMMENTED** [review]
Nice work of flushing all of these out! I only have a couple of nits.

> Reviewed 6 of 6 files at r7, 1 of 1 files at r8, 1 of 1 files at r9, 1 of 1 files at r10, 4 of 4 files at r11, 2 of 2 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108566#-:-Nbbf-UxE4evhHecH_XD:b-e2gc73)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[`pkg/cli/democluster/demo_cluster.go` line 1014 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/108566#-NbbcBVlCwNbbvJ0s_99:-NbbcBVlCwNbbvJ0s_9A:b9jvx60) ([raw file](https://github.com/cockroachdb/cockroach/blob/649779f9f9e7e9d886e208136af69f28da3b4947/pkg/cli/democluster/demo_cluster.go#L1014)):*
> ```Go
> 		if s.nodeID == nodeID {
> 			serverIdx = i
> 			break
> ```

nit: why not return here right away once match is found?
___
*[`pkg/cli/democluster/demo_cluster.go` line 1030 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/108566#-NbbcHed-jEJhSlhPOZU:-NbbcHed-jEJhSlhPOZV:b-y86afw) ([raw file](https://github.com/cockroachdb/cockroach/blob/649779f9f9e7e9d886e208136af69f28da3b4947/pkg/cli/democluster/demo_cluster.go#L1030)):*
> ```Go
> 	serverIdx := c.findServer(roachpb.NodeID(nodeID))
> 
> 	if serverIdx == -1 {
> ```

nit: consider extracting `-1` as a unexported global constant since it's used in several spots.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108566)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[`pkg/cli/democluster/demo_cluster.go` line 1014 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/108566#-NbbcBVlCwNbbvJ0s_99:-NbpuglR4zl_z8ofqdah:b-26aly6) ([raw file](https://github.com/cockroachdb/cockroach/blob/649779f9f9e7e9d886e208136af69f28da3b4947/pkg/cli/democluster/demo_cluster.go#L1014)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: why not return here right away once match is found?
</blockquote></details>

Done in a later commit.
___
*[`pkg/cli/democluster/demo_cluster.go` line 1030 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/108566#-NbbcHed-jEJhSlhPOZU:-Nbpuigq3ez9YFHdRIUc:bar7kkm) ([raw file](https://github.com/cockroachdb/cockroach/blob/649779f9f9e7e9d886e208136af69f28da3b4947/pkg/cli/democluster/demo_cluster.go#L1030)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: consider extracting `-1` as a unexported global constant since it's used in several spots.
</blockquote></details>

Unneeded in the later commit.

**Pull request merged**